### PR TITLE
Adding IdGen to LogicalPlans

### PR DESCRIPTION
**Pull request by sherfert**
First part of cool and shiny attribution of trees.

### CODE REVIEW

File: community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  val array:ArrayBuffer[Unchangeable[T]] = new ArrayBuffer[Unchangeable[T]]()
+
+  def set(id:Id, t:T): Unit = {
+    val requiredSize = id.x + 1
+    if (array.size < requiredSize) {
+      while (array.size < requiredSize)
+        array += new Unchangeable
+      array(id.x).value = t
+    } else {
+      val prev = array(id.x)
```
Comment (systay) [inline community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala]: never used

File: community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+import scala.collection.mutable.ArrayBuffer
+
+trait Attribute[T] {
+
+  // TODO crappy perf?
+  val array:ArrayBuffer[Unchangeable[T]] = new ArrayBuffer[Unchangeable[T]]()
+
+  def set(id:Id, t:T): Unit = {
+    val requiredSize = id.x + 1
+    if (array.size < requiredSize) {
```
Comment (systay) [inline community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala]: Could we move this to its own method?

**Comment by Lojjs** [comment]
Two of the tests in MatchAcceptanceTest are not working right now, need a bug fix for Slotted pipes. A lot of other tests are also broken.

File: community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala (hunk 2 above)
Comment (Lojjs) [reply community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala]: @systay which part do you mean?

**Comment by sherfert** [comment]
Currently fails in GrammarStressIT. To reproduce:

```
test("oh no") {
    assertQuery("MATCH (:L1)-[:T1 *..1 {p1: 1}]->(n2 :L2 {p2: 7})-[r2s :T2 *..2]->(:L3) RETURN 42")
}
```

It seems the SlotAllocation for VarExpand got somehow affected. @fickludd Can you help?

File: community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala (hunk 2 above)
Comment (systay) [reply community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala]: the array resizing

File: community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/LogicalPlan.scala [hunk 3]
```diff
@@ -37,55 +38,32 @@ non-leaf nodes are algebraic operators like selections, projections, and joins.
...
+abstract class LogicalPlan(idGen: IdGen)
   extends Product
   with Foldable
   with Strictness
-  with RewritableWithMemory {
+  with Rewritable {
 
   self =>
 
+  // FIXME this should be removed ASAP
```
Comment (systay) [inline community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/LogicalPlan.scala]: Maybe add to the comment that fact that the bug has been reported and fixed, and when we can expect to be able to remove it?

File: community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala (hunk 2 above)
Comment (Lojjs) [reply community/cypher/util-3.4/src/main/scala/org/neo4j/cypher/internal/util/v3_4/attribution/Attribute.scala]: Okey, have fixed!

File: community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/LogicalPlan.scala (hunk 3 above)
Comment (Lojjs) [reply community/cypher/cypher-logical-plans-3.4/src/main/scala/org/neo4j/cypher/internal/v3_4/logical/plans/LogicalPlan.scala]: Comment improved

**Review by fickludd — APPROVED** [review]
Reviewed this PR by accident when looking at #10710. My concerns were addressed there (or maybe here) so this is good to go!

**Review by systay — APPROVED** [review]
Waiting for green build

**Pull request merged**